include/deal.II/fe/fe_enriched.h

```diff
@@ -0,0 +1,545 @@
+  //@}
+
+
+  /**
+   * Return enrichment functions
+   */
+  const std::vector<std::vector<std::function<const Function<dim> *(const typename Triangulation<dim, spacedim>::cell_iterator &) > > >
+  get_enrichments() const;
+
+  /**
+   * Return the underlying FESystem object.
+   */
+  const FESystem<dim,spacedim> &
+  get_fe_system() const;
+
+protected:
+
+  /**
+   * For each Finite Element (base number) and each enrichment function (base_index)
+   * this struct will contain values, gradients and hessians of the enrichment functions.
+   */
+  struct EnrichmentValues
+  {
+    std::vector<double> values;
+    std::vector<Tensor<1,dim> > gradients;
+    std::vector<SymmetricTensor<2, dim> > hessians;
+  };
+
+
+  /**
+   * A class to hold internal data needed for evaluation of this FE at quadrature points.
+   */
+  class InternalData : public FiniteElement<dim,spacedim>::InternalDataBase
+  {
+  public:
+    /**
+     * Constructor. Is called by the @p get_data function.
+     * As we aggregate FESystem's internal data, we only call it after
+     * the FESystem::get_data() is called.
+     */
+    InternalData ( typename FESystem<dim,spacedim>::InternalData *fesystem_data);
+
+    /**
+     * Destructor. Deletes @p fesystem_data pointer.
+     */
+    ~InternalData();
+
+    /**
+     * Gives read-access to the pointer to a @p InternalData of the @p
+     * <code>base_no</code>th base element of FESystem's data.
+     */
+    typename FiniteElement<dim,spacedim>::InternalDataBase &
+    get_fe_data (const unsigned int base_no) const;
+
+    /**
+     * Gives read-access to the pointer to an object to which into which the
+     * <code>base_no</code>th base element will write its output when calling
+     * FiniteElement::fill_fe_values() and similar functions.
+     */
+    internal::FEValues::FiniteElementRelatedData<dim,spacedim> &
+    get_fe_output_object (const unsigned int base_no) const;
+
+    /**
+     * Aggregate FESystem's internal data. It is used it every time
+     * we call FESystem's fill_fe_values() and alike.
+     */
+    typename FESystem<dim,spacedim>::InternalData *fesystem_data;
+
+    /**
+     * For each FE used in enrichment (base number <code>i</code>) and each enrichment function
+     * (base multiplicity <code>j</code>), <code>enrichment_values[i][j]</code> will be used to store
+     * possibly requested values, gradients and hessians of enrichment function <code>j</code>.
+     *
+     * The variable is made mutable as InternalData's provided to fill_fe_values and alike
+     * are const.
+     *
+     * @note We do not want to store this information in the finite element object itself,
+     * because this would mean that (i) only one FEValues object could use a finite element object at a time,
+     * and (ii) that these objects could not be used in a multithreaded context.
+     */
+    mutable std::vector<std::vector<EnrichmentValues> > enrichment;
```

Comment:
i don't think so because each FE used with enrichment can be coupled with an arbitrary number of enrichment functions. That is, you may have
```
FE_Q(2)  // base FE
FE_Q(1)  + sin(x) + cos(x) // first element with enrichment
FE_Q(2) + exp(-x) // second element with enrichment
```